vignettes/corrected-se.Rmd

```diff
@@ -0,0 +1,322 @@
+
+where $V_{\gamma}$ is the naive covariance matrix of the structural parameter estimates $\hat{\boldsymbol{\gamma}}$ assuming the measurement error variance parameter, $\boldsymbol{\theta}$, is known, $\boldsymbol{H}_\gamma$ is the Hessian matrix of the log-likelihood $\ell$ with respect to $\hat{\boldsymbol{\gamma}}$, and $V_{\theta}$ can be obtained in the first-stage measurement model analysis.
+
+```{r load-pkg, message = FALSE}
+library(lavaan)
+library(R2spa)
+library(numDeriv)
+library(boot)
+```
+
+This example is based on the Political Democracy data used in <https://lavaan.ugent.be/tutorial/sem.html> and as an example in Bollen (1989)'s book.
+
+## Separate Measurement Models
+
+```{r}
+model <- '
+  # latent variable definitions
+     ind60 =~ x1 + x2 + x3
+     dem60 =~ y1 + y2 + y3 + y4
+
+  # regressions
+    dem60 ~ ind60
+'
+```
+
+```{r}
+cfa_ind60 <- cfa("ind60 =~ x1 + x2 + x3", data = PoliticalDemocracy)
+# Regression factor scores
+fs1 <- get_fs_lavaan(cfa_ind60, vfsLT = TRUE)
+# Delta method variance of (loading, intercept)
+vldev1 <- attr(fs1, which = "vfsLT")
+cfa_dem60 <- cfa("dem60 =~ y1 + y2 + y3 + y4",
+                 data = PoliticalDemocracy)
+# Regression factor scores
+fs2 <- get_fs_lavaan(cfa_dem60, vfsLT = TRUE)
+# Delta method variance of (loading, intercept)
+vldev2 <- attr(fs2, which = "vfsLT")
+fs_dat <- data.frame(
+  fs_ind60 = fs1$fs_ind60,
+  fs_dem60 = fs2$fs_dem60
+)
+# Combine sampling variance of loading and error variance
+# Note: loadings first, then error variance
+vldev <- block_diag(vldev1, vldev2)[c(1, 3, 2, 4), c(1, 3, 2, 4)]
+```
+
+```{r}
+# 2S-PA
+# Assemble loadings
+ld <- block_diag(attr(fs1, which = "fsL"),
+                 attr(fs2, which = "fsL"))
+ev <- block_diag(attr(fs1, which = "fsT"),
+                 attr(fs2, which = "fsT"))
+tspa_fit <- tspa(model = "dem60 ~ ind60",
+                 data = fs_dat,
+                 cross_loadings = ld,
+                 vc = ev)
+# Unadjusted covariance
+vcov(tspa_fit)
+# Adjusted covariance matrix
+(vc_cor <- vcov_corrected(tspa_fit, vfsLT = vldev, which_free = c(1, 4, 5, 7)))
+# Corrected standard errors
+sqrt(diag(vc_cor))
+```
+
+Standardized solution
+
+```{r}
+tspa_est_std <- function(ld_ev) {
+  ld1 <- ld
+  ev1 <- ev
+  ld1[c(1, 4)] <- ld_ev[1:2]
+  ev1[c(1, 4)] <- ld_ev[3:4]
+  tfit <- tspa(model = "dem60 ~ ind60",
+               data = fs_dat,
+               cross_loadings = ld1,
+               vc = ev1)
+  standardizedSolution(tfit)$est.std[8]
+}
+tspa_est_std(c(ld[c(1, 4)], ev[c(1, 4)]))
+# Jacobian
+jac_std <- numDeriv::jacobian(tspa_est_std,
+                              x = c(ld[c(1, 4)], ev[c(1, 4)]))
+# Corrected standard error
+sqrt(
+  standardizedSolution(tspa_fit)[8, "se"]^2 +
+    jac_std %*% vldev %*% t(jac_std)
+)
+```
+
+Compare to joint model
```

Comment:
Minor suggestion: "Compare to joint structural and measurement model"? because we have "Joint measurement model" later